### PR TITLE
[WIP] Add map_blocks.

### PR DESCRIPTION
**Pull request by dcherian**
- Related #3147 
 - [ ] Tests added
 - [x] Passes `black . && mypy . && flake8`
 - [ ] Fully documented, including `whats-new.rst` for all changes and `api.rst` for new API

ping @mrocklin @sofroniewn @shanaxel42

### CODE REVIEW

File: xarray/core/parallel.py [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+            slices.append(slice(start, stop))
+            start = stop
+        chunk_slices[dim] = slices
+
+    return chunk_slices
+
+
+def map_blocks(func, darray):
+    """
+    A version of dask's map_blocks for DataArrays.
```
Comment (mrocklin) [inline xarray/core/parallel.py]: We should maybe not assume that the user is familiar with Dask?  I don't know how common this is among Xarray users.

```suggestion
    Apply a function across each chunk of the DataArray in parallel
    
    This is similar to Dask Array's map_blocks function, but your function 
    receives a DataArray rather than a Numpy Array.
```

File: xarray/core/parallel.py [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+        if not isinstance(result, type(darray)):
+            raise ValueError("Result is not the same type as input.")
+        if result.shape != darray.shape:
+            raise ValueError("Result does not have the same shape as input.")
+        return result
+
+    meta_array = DataArray(darray.data._meta, dims=darray.dims)
+    result_meta = func(meta_array)
+
+    name = "%s-%s" % (darray.name or func.__name__, dask.base.tokenize(darray))
```
Comment (mrocklin) [inline xarray/core/parallel.py]: You might want to use `dask.utils.funcname` in order to get the function name.  This handles some additional cases we've seen, like `functools.partial`.

Comment (mrocklin) [reply xarray/core/parallel.py]: What is `darray.name`?  It might be best to use the function name regardless.  This is the name that shows up in things like the progress bars in the dashboard.

File: xarray/core/parallel.py [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+        for chunk in chunks:
+            stop = start + chunk
+            slices.append(slice(start, stop))
+            start = stop
+        chunk_slices[dim] = slices
+
+    return chunk_slices
+
+
+def map_blocks(func, darray):
```
Comment (mrocklin) [inline xarray/core/parallel.py]: We might consider accepting `*args, **kwargs` here, at least in the simple case when they aren't other Dask/Xarray collections.  That would allow for simple things like ...

```python
darray.map_blocks(operator.add, 10)
```

So maybe something like ...

```python
for arg in args + tuple(kwargs.values()):
    if dask.base.is_dask_collection(arg):
        raise NotImplementedError(...)
```

And then use the task`(_wrapper, func, (DataArray, key, ...), args, kwargs)`

File: xarray/core/parallel.py [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+    def _wrapper(darray):
+        result = func(darray)
+        if not isinstance(result, type(darray)):
+            raise ValueError("Result is not the same type as input.")
+        if result.shape != darray.shape:
+            raise ValueError("Result does not have the same shape as input.")
+        return result
+
+    meta_array = DataArray(darray.data._meta, dims=darray.dims)
+    result_meta = func(meta_array)
```
Comment (mrocklin) [inline xarray/core/parallel.py]: We might want to have a try-except around this, and raise some informative error or warning if it doesn't work.

**Comment by crusaderky** [comment]
Hi,

A few design opinions:

1. could we call it just "map"? It makes sense as this thing would be very useful for non-dask based arrays too. Working routinely with scipy (chiefly with scipy.stats transforms), I tire a lot of writing very verbose ``xarray.apply_ufunc`` calls.

2. could we have it as a method of DataArray and Dataset, to allow for method chaining?

e.g.
```python
myarray.map(func1).chunk().map(func2).sum().compute()
```

File: xarray/core/parallel.py (hunk 3 above)
Comment (crusaderky) [reply xarray/core/parallel.py]: +1 for ``*args, **kwargs``

Comment (crusaderky) [inline xarray/core/parallel.py]: Can we expose dtype=None as in the underlying dask function? It would save some CPU when

File: xarray/core/parallel.py [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+                {
+                    dim_name: darray[dim_name][slicers[dim_name][index]]
+                    for dim_name, index in zip(darray.dims, key[1:])
+                },
+                darray.dims,
+            ),
+        )
+        for key in dask_keys
+    }
+
```
Comment (crusaderky) [inline xarray/core/parallel.py]: Instead of building a graph by hand, why aren't you simply calling ``dask.array.map_blocks``?

**Comment by crusaderky** [comment]
@shoyer let me rephrase it - apply_ufunc is extremely powerful, and when you need to cope with all possible shape transformations, I suspect its verbosity is quite necessary.
It's just that, when all you need to do is apply an elementwise, embarassingly parallel function (80% of the times in my real life experience), apply_ufunc is overkill.

The thing I have against the name map_blocks is that backends other than dask have no notion of blocks...

**Comment by dcherian** [comment]
I started prototyping a Dataset version. Here's what I have:

``` python
import dask
import xarray as xr

darray = xr.DataArray(np.ones((10, 20)), 
                  dims=['x', 'y'], 
                  coords={'x': np.arange(10), 'y': np.arange(100, 120)})
dset = darray.to_dataset(name='a')
dset['b'] = dset.a + 50
dset['c'] = (dset.x + 20)
dset = dset.chunk({'x': 4, 'y': 5})
```

The function I'm applying takes a dataset and returns a DataArray because that's easy to test without figuring out how to assemble everything back into a dataset.
``` python
import itertools

# function takes dataset and returns dataarray so that I can check that things work without reconstructing a dataset
def function(ds):
    return ds.a + 10

dataset_dims = list(dset.dims)

graph = {}
gname = 'dsnew'

# map dims to list of chunk indexes
# If different variables have different chunking along the same dim
# the call to .chunks will raise an error.
ichunk = {dim: range(len(dset.chunks[dim])) for dim in dataset_dims}

# iterate over all possible chunk combinations
for v in itertools.product(*ichunk.values()):
    chunk_index_dict = dict(zip(dataset_dims, v))
    data_vars = {}
    for name, variable in dset.data_vars.items():
        # why do does dask_keys have an extra level?
        # the [0] is not required for dataarrays
        var_dask_keys = variable.__dask_keys__()[0]
        
        # recursively index into dask_keys nested list
        chunk = var_dask_keys
        for dim in variable.dims:
            chunk = chunk[chunk_index_dict[dim]]
            
        # I have key corresponding to chunk
        # this tuple is in a dictionary passed to xr.Dataset()
        # dask doesn't seem to replace this with a numpy array at execution time.
        data_vars[name] = (variable.dims, chunk)
        
    graph[(gname, ) + v] = (function, (xr.Dataset, data_vars))

final_graph = dask.highlevelgraph.HighLevelGraph.from_collections(name, graph, dependencies=[dset])
```

Elements of the graph look like
```
('dsnew', 0, 0): (<function __main__.function(ds)>,
  (xarray.core.dataset.Dataset,
   {'a': (('x', 'y'), ('xarray-a-f178df193efafa67203f3862b3f9f0f4', 0, 0)),
    'b': (('x', 'y'), ('xarray-b-e2d8d06bb9e5c1f351671a94816bd331', 0, 0)),
    'c': (('x',), ('xarray-c-d90f8b2af715b53f4c170be391239655', 0))}))
```

This doesn't work because dask doesn't replace the keys by numpy arrays when the `xr.Dataset` call is executed.  The graph is "disconnected": 
image


I'm not sure what I'm doing wrong here. An equivalent version for DataArrays works perfectly.

**Comment by mrocklin** [comment]
Dask doesn't traverse through tuples to find possible keys, so the keys here are hidden from view:

```python
   {'a': (('x', 'y'), ('xarray-a-f178df193efafa67203f3862b3f9f0f4', 0, 0)),
```

I recommend changing wrapping tuples with lists:

```diff
-   {'a': (('x', 'y'), ('xarray-a-f178df193efafa67203f3862b3f9f0f4', 0, 0)),
+   {'a': [('x', 'y'), ('xarray-a-f178df193efafa67203f3862b3f9f0f4', 0, 0)],
```

**Comment by dcherian** [comment]
Thanks @mrocklin. Unfortunately that doesn't work with the Dataset constructor. With a list it treats it as array-like

```
    The following notations are accepted:

    - mapping {var name: DataArray}
    - mapping {var name: Variable}
    - mapping {var name: (dimension name, array-like)}
    - mapping {var name: (tuple of dimension names, array-like)}
    - mapping {dimension name: array-like}
      (it will be automatically moved to coords, see below)
```

Unless @shoyer has another idea, I guess I can insert creating a DataArray into the graph and then refer to those keys in the Dataset constructor.

**Comment by mrocklin** [comment]
Then you can construct a tuple as a task `((1, 2, 3),)` -> `(tuple, [1, 2, 3])`

**Comment by dcherian** [comment]
Thanks. That worked. I have a new version up in #3276 that works with both DataArrays and Datasets.

**Comment by mrocklin** [comment]
I'm glad to see progress here.  FWIW, I think that many people would be quite happy with a version that just worked for DataArrays, in case that's faster to get in than the full solution with DataSets.

**Comment by dcherian** [comment]
Closing in favour of #3276
